### PR TITLE
603 kingfisher process middleware ng

### PR DESCRIPTION
**Pull request by jakubkrafka**
First version of kingfisher NG integration.

### CODE REVIEW

File: kingfisher_scrapy/extensions.py [hunk 1]
```diff
@@ -288,3 +289,101 @@ def from_crawler(cls, crawler):
...
+            "check": True,
+        }
+
+        response = self._post("api/v1/create_collection", data)
+
+        if not response.ok:
+            spider.logger.warning(
+                'Failed to POST create_collection. API status code: {}'.format(response.status_code))
+        else:
+            response_data = json.loads(response.text)
```
Comment (jpmckinney) [inline kingfisher_scrapy/extensions.py:343]:
```suggestion
            response_data = response.json()
```

File: kingfisher_scrapy/extensions.py [hunk 2]
```diff
@@ -288,3 +289,101 @@ def from_crawler(cls, crawler):
...
+
+        if not response.ok:
+            spider.logger.warning(
+                'Failed to post close collection. API status code: {}'.format(response.status_code))
+
+    def item_scraped(self, item, spider):
+        """
+        Sends an API request to store the file in Kingfisher Process.
+        """
+        if not item.get('post_to_api', True) or isinstance(item, PluckedItem):
```
Comment (jpmckinney) [inline kingfisher_scrapy/extensions.py:364]: `post_to_api` has been gone a long time now, I think.

```suggestion
        if if isinstance(item, PluckedItem):
```

File: kingfisher_scrapy/extensions.py [hunk 3]
```diff
@@ -288,3 +289,101 @@ def from_crawler(cls, crawler):
...
+        response = self._post("api/v1/create_collection", data)
+
+        if not response.ok:
+            spider.logger.warning(
+                'Failed to POST create_collection. API status code: {}'.format(response.status_code))
+        else:
+            response_data = json.loads(response.text)
+            self.collection_id = response_data["collection_id"]
+
+    def spider_closed(self, spider, reason):
```
Comment (jpmckinney) [inline kingfisher_scrapy/extensions.py:346]: We should retain the logic from `KingfisherProcessAPI`, which is currently:

```python
        # https://docs.scrapy.org/en/latest/topics/signals.html#spider-closed
        if reason not in ('finished', 'sample') or spider.pluck or spider.keep_collection_open:
            return
```

Regarding the first condition, see https://github.com/open-contracting/kingfisher-collect/issues/278 The proposal is to send the `reason` to the API.

File: kingfisher_scrapy/extensions.py [hunk 4]
```diff
@@ -288,3 +289,101 @@ def from_crawler(cls, crawler):
...
+        data = {
+            "collection_id": self.collection_id,
+            "path": os.path.join(item['files_store'], item['path'])
+        }
+
+        if isinstance(item, FileError):
+            # in case of error send info about it to api
+            data['errors'] = json.dumps(item['errors'])
+
+        response = self._post("api/v1/create_collection_file", data)
```
Comment (jpmckinney) [inline kingfisher_scrapy/extensions.py:376]: The current extension also sends the `file_name` and `url` from the item, for which there are corresponding columns in the `collection_file` table.

Comment (jpmckinney) [reply kingfisher_scrapy/extensions.py:376]: Also, for FileItem's, it sends the `number`.

File: kingfisher_scrapy/extensions.py [hunk 5]
```diff
@@ -288,3 +289,101 @@ def from_crawler(cls, crawler):
         extension = cls()
         sentry_sdk.init(sentry_dsn)
         return extension
+
+
+class KingfisherProcessNGAPI:
```
Comment (jpmckinney) [inline kingfisher_scrapy/extensions.py:294]: The current extension also implements `item_error` and `spider_error` signals.

**Comment by jpmckinney** [comment]
Also, ideally, we would have coverage for the error scenarios. The existing tests provide a lot of inspiration for how to construct new tests.

File: kingfisher_scrapy/extensions.py [hunk 6]
```diff
@@ -288,3 +289,101 @@ def from_crawler(cls, crawler):
...
+    @classmethod
+    def from_crawler(cls, crawler):
+        url = crawler.settings['KINGFISHER_NG_API_URI']
+        username = crawler.settings['KINGFISHER_NG_API_USERNAME']
+        password = crawler.settings['KINGFISHER_NG_API_PASSWORD']
+
+        if not url:
+            raise NotConfigured('KINGFISHER_NG_API_URI is not set.')
+
+        if (username and not password) or (password and not username):
```
[the remark this answers is not in the document]
Comment (jakubkrafka) [reply kingfisher_scrapy/extensions.py:313]: Of course, it's XOR indeed, but my preference is the more "readable" variant. But, no strong reasoning behind it, just a matter of taste.

File: kingfisher_scrapy/extensions.py [hunk 7]
```diff
@@ -288,3 +289,101 @@ def from_crawler(cls, crawler):
         extension = cls()
         sentry_sdk.init(sentry_dsn)
         return extension
+
+
+class KingfisherProcessNGAPI:
+    """
+    If the ``KINGFISHER_API_URI`` and ``KINGFISHER_API_KEY`` environment variables or configuration settings are set,
```
[the remark this answers is not in the document]
Comment (jakubkrafka) [reply kingfisher_scrapy/extensions.py:296]: Okay, URL is preferred on my side as well (actually, i found a bug in the code, where URL was used instead URI), just wanted to be consistent.

File: kingfisher_scrapy/extensions.py (hunk 3 above)
Comment (jakubkrafka) [reply kingfisher_scrapy/extensions.py:347]: Implemented - the reason will be saved as a note.

Comment (jakubkrafka) [reply kingfisher_scrapy/extensions.py:348]: The reason is sent to API and processed/saved as a collection note.

File: kingfisher_scrapy/extensions.py [hunk 8]
```diff
@@ -3,6 +3,7 @@
 import json
 import os
 
+import requests
```
[the remark this answers is not in the document]
Comment (jakubkrafka) [reply kingfisher_scrapy/extensions.py:6]: Added treq, makes sense in most places. Except for the first call to create a collection. The collection id is used in the next steps (create file etc.). We need to stick to synchronous call here. Unfortunately, I don't see a nice simple solution here:(

File: kingfisher_scrapy/extensions.py (hunk 5 above)
Comment (jakubkrafka) [reply kingfisher_scrapy/extensions.py:295]: That's true but it seems to me, that those methods are never used. At least not connected to any signals in the from_crawler method. But I'm not an expert in scrapy extensions - is there some magic auto-connect?

**Comment by jpmckinney** [comment]
@jakubkrafka I also thought Kingfisher Collect would directly enqueue files in RabbitMQ rather than sending files to Kingfisher Process. Right now we're having issues with Kingfisher Process not being able to handle the volume of requests it receives. I had a look and the new version also interacts with the database in the web request (I thought we were going to enqueue individual file loads). Interacting with the database in a web request is likely to cause the same issues as we have now.

File: kingfisher_scrapy/extensions.py (hunk 5 above)
Comment (jpmckinney) [reply kingfisher_scrapy/extensions.py:295]: Indeed, that's a bug! https://github.com/open-contracting/kingfisher-collect/issues/659

File: kingfisher_scrapy/extensions.py [hunk 9]
```diff
@@ -293,8 +294,9 @@ def from_crawler(cls, crawler):
 
 class KingfisherProcessNGAPI:
     """
-    If the ``KINGFISHER_API_URI`` and ``KINGFISHER_API_KEY`` environment variables or configuration settings are set,
+    If the ``KINGFISHER_NG_API_URL`` environment variable or configuration setting is set,
     then messages are sent to a Kingfisher Process API for the ``item_scraped`` and ``spider_closed`` signals.
+
```
Comment (jpmckinney) [inline kingfisher_scrapy/extensions.py:299]: Remove extra newline.

File: kingfisher_scrapy/extensions.py [hunk 10]
```diff
@@ -373,17 +381,29 @@ def item_scraped(self, item, spider):
...
-        if not response.ok:
-            spider.logger.warning(
-                'Failed to POST create_collection_file. API status code: {}'.format(response.status_code))
+        self._post_async("api/v1/create_collection_file", data, response_callback)
 
-    def _post(self, url, data):
+    def _post_sync(self, url, data):
         """
-        Wrapper around the requests. Add auth if necessary.
+        Wrapper around the requests to api. Adds auth if necessary.
```
Comment (jpmckinney) [inline kingfisher_scrapy/extensions.py:393]:
```suggestion
        Posts synchronous requests to Kingfisher Process' API, adding authentication if needed.
```

File: kingfisher_scrapy/extensions.py [hunk 11]
```diff
@@ -373,17 +381,29 @@ def item_scraped(self, item, spider):
...
         """
         if self.username and self.password:
             return requests.post("{}/{}".format(self.url, url), json=data, auth=(self.username, self.password))
         else:
             return requests.post("{}/{}".format(self.url, url), json=data)
+
+    def _post_async(self, url, data, callback):
+        """
+        Wrapper around the treq to api. Adds auth if necessary.
```
Comment (jpmckinney) [inline kingfisher_scrapy/extensions.py:402]:
```suggestion
        Posts asynchronous requests to Kingfisher Process' API, adding authentication if needed.
```

File: kingfisher_scrapy/extensions.py [hunk 12]
```diff
@@ -373,17 +381,29 @@ def item_scraped(self, item, spider):
...
+        """
+        if self.username and self.password:
+            request = treq.post("{}/{}".format(self.url, url), json=data, auth=(self.username, self.password))
+        else:
+            request = treq.post("{}/{}".format(self.url, url), json=data)
```
Comment (jpmckinney) [inline kingfisher_scrapy/extensions.py:407]: Reduce repetition. Can do same for `_post_sync`.

```suggestion
        kwargs = {}
        if self.username and self.password:
            kwargs['auth'] = (self.username, self.password)
        request = treq.post("{}/{}".format(self.url, url), json=data, **kwargs)
```

File: requirements.in [hunk 13]
```diff
@@ -11,4 +11,4 @@ scrapyd-client
 sentry-sdk
 treq
 twisted
-requests
+requests
```
Comment (jpmckinney) [inline requirements.in:14]: Keep newline at end of file.

File: tests/extensions/test_kingfisher_process_api_NG.py [hunk 14]
```diff
@@ -28,19 +28,19 @@ def test_from_crawler():
 
 def test_from_crawler_missing_uri():
     spider = spider_with_crawler(settings={
-        "KINGFISHER_API_URI": "missign",
+        "KINGFISHER_API_URL": "missign",
```
Comment (jpmckinney) [inline tests/extensions/test_kingfisher_process_api_NG.py:31]:
```suggestion
        "KINGFISHER_API_URL": "missing",
```

File: kingfisher_scrapy/extensions.py (hunk 12 above)
Comment (jakubkrafka) [reply kingfisher_scrapy/extensions.py:408]: Frankly, not a big fan of kwargs here. In 1:1 cases, I prefer readability before duplication/repetition. But you are the codebase master.

**Comment by jakubkrafka** [comment]
> @jakubkrafka I also thought Kingfisher Collect would directly enqueue files to load in RabbitMQ rather than sending files to Kingfisher Process's web API. Right now we're having issues with Kingfisher Process not being able to handle the volume of requests it receives. ...

What are the expectations here actually? With a very basic perf test, on dqt01 server (AWS, 4CPU, slow disk), with no special settings, I'm able to handover about 270 files per second. 
Note:The file is not processed after this call, it's only planned. The actual load happens in a later phase.

**Comment by jpmckinney** [comment]
Restarted in #745

**Pull request closed without merging**